producer-c/producer-cloudwatch-integ/KvsVideoOnlyStreamingSampleCloudwatch.cpp

```diff
@@ -0,0 +1,228 @@
+#define DEFAULT_FPS_VALUE                   25
+#define DEFAULT_STREAM_DURATION             (20 * HUNDREDS_OF_NANOS_IN_A_SECOND)
+
+#define NUMBER_OF_FRAME_FILES               403
+#define CANARY_METADATA_SIZE                (SIZEOF(INT64) + SIZEOF(UINT32) + SIZEOF(UINT32) + SIZEOF(UINT64))
+
+#define CANARY_FILE_LOGGING_BUFFER_SIZE     (200 * 1024)
+#define CANARY_MAX_NUMBER_OF_LOG_FILES      10
+VOID sigintHandler(INT32 sigNum)
+{
+	ATOMIC_STORE_BOOL(&sigCaptureInterrupt, TRUE);
+}
+
+// add frame pts, frame index, original frame size, CRC to beginning of buffer
+VOID addCanaryMetadataToFrameData(PFrame pFrame) {
+    putInt64((PINT64)pFrame->frameData, pFrame->presentationTs / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+    putInt32((PINT32) (pFrame->frameData + SIZEOF(UINT64)), pFrame->index);
+    putInt32((PINT32) (pFrame->frameData + SIZEOF(UINT64) + SIZEOF(UINT32)), pFrame->size);
+    MEMSET(pFrame->frameData + SIZEOF(UINT64) + SIZEOF(UINT32) + SIZEOF(UINT32), 0x00, SIZEOF(UINT64));
+    putInt64((PINT64) (pFrame->frameData + SIZEOF(UINT64) + SIZEOF(UINT32) + SIZEOF(UINT32)), COMPUTE_CRC32(pFrame->frameData, pFrame->size));
+}
+
+VOID createCanaryFrameData(PFrame pFrame) {
+    UINT32 i;
+
+    for (i = CANARY_METADATA_SIZE; i < pFrame->size; i++) {
+        pFrame->frameData[i] = RAND();
+    }
+    addCanaryMetadataToFrameData(pFrame);
+}
+
+PCHAR getCanaryStr(UINT32 canaryType) {
+    switch (canaryType) {
+        case 0:
+            return (PCHAR)"realtime";
+        case 1:
+            return (PCHAR)"offline";
+        default:
+            return (PCHAR)"";
+    }
+}
+
+VOID adjustStreamInfoToCanaryType(PStreamInfo pStreamInfo, UINT32 canaryType) {
+    switch (canaryType) {
+        case 0:
+            pStreamInfo->streamCaps.streamingType = STREAMING_TYPE_REALTIME;
+            break;
+        case 1:
+            pStreamInfo->streamCaps.streamingType = STREAMING_TYPE_OFFLINE;
+            break;
+        default:
+            break;
+    }
+
+}
+
+INT32 main(INT32 argc, CHAR *argv[])
+{
+#ifndef _WIN32
+    signal(SIGINT, sigintHandler);
+#endif
+    SET_INSTRUMENTED_ALLOCATORS();
+    PDeviceInfo pDeviceInfo = NULL;
+    PStreamInfo pStreamInfo = NULL;
+    PStreamCallbacks pStreamCallbacks = NULL;
+    PClientCallbacks pClientCallbacks = NULL;
+    CLIENT_HANDLE clientHandle = INVALID_CLIENT_HANDLE_VALUE;
+    STREAM_HANDLE streamHandle = INVALID_STREAM_HANDLE_VALUE;
+    STATUS retStatus = STATUS_SUCCESS;
+    PCHAR accessKey = NULL, secretKey = NULL, sessionToken = NULL, streamNamePrefix = NULL, canaryTypeStr = NULL, region = NULL, cacertPath = NULL;
+    CHAR streamName[MAX_STREAM_NAME_LEN + 1];
+    Frame frame;
+    UINT32 frameIndex = 0, fileIndex = 0, canaryType = 0;
+    UINT64 fragmentSizeInByte = 0;
+    UINT64 lastKeyFrameTimestamp = 0;
+    PCanaryStreamCallbacks pCanaryStreamCallbacks = NULL;
+    pCanaryStreamCallbacks = (PCanaryStreamCallbacks) MEMCALLOC(1, SIZEOF(CanaryStreamCallbacks));
```

Comment:
As discussed, the right pattern in this case is to let the createCallbacksBlah function to own the allocation and the creation of the actual object.

Reply:
Done